src/models/MusicPlayer.ts

```diff
@@ -0,0 +1,164 @@
+    readonly client: DiscordClient;
+
+    readonly guildId: string;
+
+    channelId: string | undefined;
+
+    voiceChannelId: string | undefined;
+
+    tracks: YouTubeVideo[] = [];
+
+    // Track position index
+    trackAt: number = 0;
+
+    loopMode: LoopMode = 'off';
+
+    quality: StreamQuality = StreamQuality.lowest;
+
+    idleTimer = 60000;
+
+    private connection: VoiceConnection | undefined;
+
+    private stopCalled = false;
+
+    private timeout: NodeJS.Timeout | undefined;
+
+    constructor(client: DiscordClient, guildId: string) {
+        super({ behaviors: { noSubscriber: NoSubscriberBehavior.Play } });
+        this.client = client;
+        this.guildId = guildId;
+    }
+
+    connect = (voiceChannel: VoiceChannel) => {
+        // Dedupe
+        if (this.voiceChannelId && this.voiceChannelId === voiceChannel.id)
+            return;
+        this.voiceChannelId = voiceChannel.id;
+        this.connection = joinVoiceChannel({
+            guildId: this.guildId,
+            channelId: this.voiceChannelId,
+            adapterCreator: voiceChannel.guild.voiceAdapterCreator,
+        });
+        this.connection.subscribe(this);
+
+        this.setTimeout();
+    };
+
+    disconnect = (): void => {
+        this.connection?.destroy();
+        this.connection = undefined;
+        this.voiceChannelId = undefined;
+    };
+
+    add = async (track: YouTubeVideo) => {
+        this.tracks.push(track);
+
+        // If we're not playing anything, start playing.
+        if (
+            this.tracks.length === 1 ||
+            this.state.status === AudioPlayerStatus.Idle
+        ) {
+            this.trackAt = this.tracks.length;
+            await this.play(track);
+        }
+
+        // Probably not strictly necessary to return tracklength.
+        return this.tracks.length;
+    };
+
+    remove = (trackNumber: number) => {
+        if (trackNumber < 0) return;
+
+        // Get currently playing, if applicable
+        let track: YouTubeVideo;
+        if (this.trackAt !== 0) {
+            track = this.tracks[this.trackAt - 1];
+        }
+
+        this.tracks.splice(trackNumber - 1, 1);
+
+        // Update current track position, if applicable
+        if (track) {
+            this.trackAt = this.tracks.findIndex((t) => t.id === track.id) + 1;
+        }
+    };
+
+    // Returns track data - useful for some cases where context is required
+    play = async (track: YouTubeVideo) => {
+        super.play(await createStream(track));
+        return track;
+    };
+
+    stop = (force?: boolean | undefined): boolean => {
+        this.stopCalled = true;
+        return super.stop(force);
+    };
+
+    // Attempts to parse mode string
+    setLoop = (mode: string) => {
+        this.loopMode = (mode as LoopMode) || 'off';
+    };
+
+    setQuality = (quality: number) => {
+        // TODO: maybe do some rounding or something
+        if (!Number.isInteger(quality)) return;
```

Comment:
suggest logging the value was not an integer.
Throwing an error might also make sense depending on how this method is used.

Reply:
yeah this is kind of a wip. at the moment no way to adjust this. as above, thinking this will be some kind of user input, will need to treat it as string THEN parse it to make sure it's safe